### PR TITLE
MGMT-8202: recalc suggested roles on host/operator changes

### PR DESCRIPTION
**Pull request by slaviered**
Trigger recalculation of suggested role for auto-assign hosts
when hosts are added or deleted and when operators are added
or deleted.

These triggers affect the considerations for role assignment
so reevaluation is in order.

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @
/cc @

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Review by avishayt — APPROVED** [review]

File: internal/bminventory/inventory.go [hunk 1]
```diff
@@ -4365,6 +4376,21 @@ func (b *bareMetalInventory) V2RegisterHost(ctx context.Context, params installe
...
+	//re-calculation occurs only when the host is bound to a multi-node
+	//day 1 cluster and more hosts are present.
+	if cluster != nil && triggerAutoAssign {
+		updated, as_err := b.clusterApi.ResetAutoAssignRoles(ctx, cluster, transaction.AddForUpdateQueryOption(tx))
+		if as_err != nil {
+			return common.GenerateErrorResponder(as_err)
+		}
+		if updated {
+			log.Infof("re-triggering auto role assignment for cluster %s following host registration", cluster.ID.String())
+			_ = b.updateHostsAndClusterStatus(ctx, cluster, tx, log)
```
Comment (filanov) [inline internal/bminventory/inventory.go:4390]: consider log in case of error

File: internal/bminventory/inventory.go [hunk 2]
```diff
@@ -4668,6 +4694,19 @@ func (b *bareMetalInventory) BindHostInternal(ctx context.Context, params instal
 		return nil, common.NewApiError(http.StatusInternalServerError, err)
 	}
 
+	//At this point the host is updated. Retrigger auto-assign role
+	//calculation since the new host may affect the assignment.
+	if cluster != nil {
+		updated, as_err := b.clusterApi.ResetAutoAssignRoles(ctx, cluster, transaction.AddForUpdateQueryOption(b.db))
```
Comment (danielerez) [inline internal/bminventory/inventory.go:4700]: maybe worth extracting the lines to a method to reuse in line 4399

File: internal/bminventory/inventory.go [hunk 3]
```diff
@@ -4381,6 +4395,24 @@ func (b *bareMetalInventory) V2RegisterHost(ctx context.Context, params installe
...
+	//calculation since the new host may affect the assignment.
+	//re-calculation occurs only when the host is bound to a multi-node
+	//day 1 cluster and more hosts are present.
+	if cluster != nil && triggerAutoAssign {
+		updated, as_err := b.clusterApi.ResetAutoAssignRoles(ctx, cluster, transaction.AddForUpdateQueryOption(tx))
+		if as_err != nil {
+			return common.GenerateErrorResponder(as_err)
+		}
+		if updated {
+			log.Infof("===> DEBUG re-triggering auto role assignment for cluster %s following host registration %s", cluster.ID.String(), hostutil.GetHostnameForMsg(host))
```
Comment (danielerez) [inline internal/bminventory/inventory.go:4408]: should it be log.Debugf instead?

File: internal/bminventory/inventory.go (hunk 2 above)
Comment (danielerez) [reply internal/bminventory/inventory.go:4707]: can't extract these lines? still look pretty similar..

File: internal/bminventory/inventory.go [hunk 4]
```diff
@@ -4669,6 +4701,19 @@ func (b *bareMetalInventory) BindHostInternal(ctx context.Context, params instal
...
+	//At this point the host is updated. Retrigger auto-assign role
+	//calculation since the new host may affect the assignment.
+	if cluster != nil {
+		updated, as_err := b.clusterApi.ResetAutoAssignRoles(ctx, cluster, transaction.AddForUpdateQueryOption(b.db))
+		if as_err != nil {
+			return nil, common.NewApiError(http.StatusInternalServerError, as_err)
+		}
+		if updated {
+			log.Infof("re-triggering auto role assignment for cluster %s following host binding", cluster.ID.String())
+			_ = b.updateHostsAndClusterStatus(ctx, cluster, b.db, log)
```
Comment (danielerez) [inline internal/bminventory/inventory.go:4713]: no need for err handling?

File: internal/cluster/cluster.go [hunk 5]
```diff
@@ -325,6 +326,25 @@ func (m *Manager) updateValidationsInDB(ctx context.Context, db *gorm.DB, c *com
 	return UpdateCluster(logutil.FromContext(ctx, m.log), db, *c.ID, *c.Status, "validations_info", string(b))
 }
 
+func (m *Manager) ResetAutoAssignRoles(ctx context.Context, cluster *common.Cluster, db *gorm.DB) (bool, error) {
+	if cluster == nil || cluster.ID == nil {
+		return false, nil
+	}
+
+	if db == nil {
```
[the remark this answers is not in the document]
Comment (danielerez) [reply internal/cluster/cluster.go:331]: still can be nil?^

File: internal/cluster/cluster.go [hunk 6]
```diff
@@ -325,6 +326,25 @@ func (m *Manager) updateValidationsInDB(ctx context.Context, db *gorm.DB, c *com
...
+	}
+
+	reply := db.Model(&models.Host{}).Where("cluster_id = ? and role = ?",
+		cluster.ID.String(), models.HostRoleAutoAssign).Update("suggested_role", models.HostRoleAutoAssign)
+
+	if err := reply.Error; err != nil {
+		m.log.WithError(err).Errorf("fail ro reset auto-assign role in cluster %s", cluster.ID.String())
+		return false, err
+	}
+	return reply.RowsAffected > 1, nil
```
[the remark this answers is not in the document]
Comment (danielerez) [reply internal/cluster/cluster.go:343]: ? ^ :)

File: internal/host/monitor.go [hunk 7]
```diff
@@ -129,7 +133,15 @@ func (m *Manager) clusterHostMonitoring() int64 {
 
 		for _, c := range clusters {
 			inventoryCache := make(InventoryCache)
-			for _, host := range sortHosts(c.Hosts) {
+			sortedHosts, canRefreshRoles := SortHosts(c.Hosts)
+			// DEBUG CODE
```
Comment (danielerez) [inline internal/host/monitor.go:137]: can remove lines 137-143 then?

File: internal/cluster/cluster.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (slaviered) [reply internal/cluster/cluster.go:331]: The scenario is the unit tests

File: internal/cluster/cluster.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (slaviered) [reply internal/cluster/cluster.go:343]: As I commented the update that need to be addresses is where the function reset other hosts (then the one that just added). So rowAffected should be 1 if the only host in the system is the current host and > 1 if some other host are reset too. in this case and only in that case a refresh is needed

File: internal/host/monitor.go (hunk 7 above)
Comment (slaviered) [reply internal/host/monitor.go:137]: some debug code is still present. uploading a fix

File: internal/bminventory/inventory.go (hunk 4 above)
Comment (slaviered) [reply internal/bminventory/inventory.go:4713]: I will log the error but I wouldn't want to fail the registration process. A subsequent refresh might fix the problem

File: subsystem/cluster_test.go [hunk 8]
```diff
@@ -3823,14 +3899,15 @@ var _ = Describe("Installation progress", func() {
 			})
 			Expect(err).NotTo(HaveOccurred())
 			c = registerClusterReply.GetPayload()
+			log.Infof("Register cluster %s", *c.ID)
```
Comment (filanov) [inline subsystem/cluster_test.go:3902]: suggesting to switch to by or remove those logs

File: internal/host/monitor.go [hunk 9]
```diff
@@ -129,7 +133,15 @@ func (m *Manager) clusterHostMonitoring() int64 {
...
+			names := make([]string, 0)
+			for _, host := range sortedHosts {
+				names = append(names, hostutil.GetHostnameForMsg(host))
+			}
+			m.log.Infof("==> DEBUG SORTED HOSTS: %v (allow refresh = %t)", names, canRefreshRoles)
```
Comment (filanov) [inline internal/host/monitor.go:142]: please remove if not needed.

Comment (slaviered) [reply internal/host/monitor.go:142]: fixed

**Pull request merged**